src/lib/middleware/pageCacheMiddleware.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  // @ts-ignore
+  PAGE_CACHE_ENABLED,
+  // @ts-ignore
+  PAGE_CACHE_TYPES,
+  // @ts-ignore
+  PAGE_CACHE_NAMESPACE,
+  // @ts-ignore
+  PAGE_CACHE_VERSION,
+  // @ts-ignore
+  PAGE_CACHE_TIMEOUT_MINUTES,
+} from "../../config"
+import { Request, Response, NextFunction } from "express"
+const cache = require("lib/cache.coffee")
+const cacheablePageTypes: string[] = PAGE_CACHE_TYPES.split("|")
+
+// Middleware will `next` and do nothing if any of the following is true:
+//
+// * page cache feature is disabled.
+// * a user is signed in.
+// * this isnt a supported cacheable path (there is an allow-list set in ENV).
+//
+// At this point, the HTML is looked up in the cache.
+// If available, the request can be responded to and ended immediately.
+// Otherwise, `next` is called and the request continues in our app.
+export const pageCacheMiddleware = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  if (!PAGE_CACHE_ENABLED) return next()
+
+  const pageType = req.path.split("/")[1]
+  if (!cacheablePath(pageType)) return next()
+
+  // @ts-ignore
+  const hasUser = !!req.user
+  if (hasUser) return next()
+
+  const handleCacheGet = (err, html) => {
+    if (err) return next()
+
+    if (html) return res.send(html)
+
+    return next()
+  }
+
+  cache.get(cacheKey(req.url), handleCacheGet)
+}
+
+// `key` should be a full URL w/ query params, and not a path.
+// This is to separate URL's like `/collect` and `/collect?acquireable=true`.
+export const cacheKey = (key: string) => {
+  return PAGE_CACHE_NAMESPACE + "-" + PAGE_CACHE_VERSION + key
+}
+
+const cacheablePath = (pageType: string) => {
+  return cacheablePageTypes.includes(pageType)
+}
+
+// `key` is a full URL w/ possibly query params
+export const cacheHtmlForPage = ({ status, key, html }) => {
+  if (!PAGE_CACHE_ENABLED || status !== 200) return
+
+  const pageTypeWithQueryParams = key.split("/")[1]
+  const pageType = pageTypeWithQueryParams.split("?")[0]
+  if (!cacheablePath(pageType)) return
+
+  const expiry = PAGE_CACHE_TIMEOUT_MINUTES * 60
```

Comment:
I'd call this `..._EXPIRY_...` since "timeout" generally means something else.

Reply:
yup, good call.